lms/djangoapps/courseware/management/commands/fix_student_module_newlines.py

```diff
@@ -0,0 +1,330 @@
+        if (ncr_grade, ncr_max_grade) != (newline_course_row.grade, newline_course_row.max_grade):
+            log.info(
+                "-> Newline course row %s has grade %s but should have grade %s",
+                newline_course_row.id,
+                (newline_course_row.grade, newline_course_row.max_grade),
+                (ncr_grade, ncr_max_grade)
+            )
+
+        if ncr_grade > ccr_grade:
+            log.info(
+                "-> Choosing newline course row %s for state and derived grade.",
+                newline_course_row.id
+            )
+            return newline_course_row.state, ncr_grade, ncr_max_grade
+
+        log.info(
+            "-> Choosing correct course row %s for state and derived grade.",
+            correct_course_row.id
+        )
+        return correct_course_row.state, ccr_grade, ccr_max_grade
+
+    def grade_for_state(self, state):
+        """Given unparsed state, return the (grade, max_grade) we should have."""
+        parsed_state = json.loads(state)
+        correct_map = parsed_state.get("correct_map")
+        if not correct_map:
+            input_state = parsed_state.get('input_state')
+            if input_state is not None:
+                return 0.0, float(len(input_state))
+            return None, None
+
+        def item_score(item):
+            """Return score for an individual correct_map entry."""
+            # Partial credit overrides all, even if correctness is 'incorrect'.
+            # We're inconsistent on whether we say the correctness is
+            # 'incorrect' or 'partially-correct' in this situation.
+            partial_points = item.get('npoints')
+            if partial_points is not None:
+                return partial_points
+
+            # Otherwise, we're either right or wrong.
+            correctness = item.get('correctness')
+            if correctness == 'correct':
+                return 1.0
+
+            return 0.0
+
+        grade = sum(item_score(item) for item in correct_map.values())
+        max_grade = float(len(correct_map))
+
+        return grade, max_grade
+
+    def row_to_keep(self, correct_course_row, newline_course_row):
+        """Determine which row's data we want to keep."""
+        # Rule 1: Take the higher grade.
+        if newline_course_row.grade > correct_course_row.grade:
+            log.info(
+                "-> Newline course record grade %s is higher than correct course record grade %s",
+                newline_course_row.grade,
+                correct_course_row.grade
+            )
+            return newline_course_row
+        elif correct_course_row.grade > newline_course_row.grade:
+            log.info(
+                "-> Correct course record grade %s is higher than newline course record grade %s",
+                correct_course_row.grade,
+                newline_course_row.grade
+            )
+            return correct_course_row
```

Comment:
True. It won't be an issue in practice for us because capa problems are the only scored type we have that turns up in the data window. We have one row of ORA2, but that stores scores in a separate table anyhow.